### PR TITLE
Moving streaming library to staging/cri-api/pkg/streaming

### LINKED ISSUE #75828: New home for pkg/kubelet/server/streaming
During the CRI-API staging/ work, pkg/kubelet/server/streaming came up as another possibility for a new home as it is being used by many CRI implementations.

- https://github.com/kubernetes/kubernetes/pull/75531#discussion_r269260936
- https://github.com/kubernetes/kubernetes/issues/75293#issuecomment-472259254

So where do we move the code from `pkg/kubelet/server/streaming`? i think i can isolate streaming/remotecommand/portforward, so next thing is to find a home for it.  choices are
- just throw that into cri-api (staging)
- new staging/ directory say cri-runtime (under kubernetes org)
- new repository under kubernetes-sigs/ named cri-runtime with history preserved

Are there other options? Do we really need to do this?

/cc @vbatts @feiskyer @derekwaynecarr @yujuhong 
/sig node

### PR DESCRIPTION
**Pull request by alok87**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
This PR is required to make life easier for people developing on cri-api, so that they do not need to vendor kubernetes for this library (as vendoring k8s is hard)
- [x] Move out streaming library to `staging/cri-api/pkg/streaming`
- [x] Isolate `remotecommand` and `portforward` dependencies
- [x] Isolate `apis/core`

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubernetes/issues/75828

**KEP under discussion**:
https://github.com/kubernetes/enhancements/pull/937

**Special notes for your reviewer**:
None
cc @dims 

**Does this PR introduce a user-facing change?**:
None

```release-note
Streaming library has a new location under staging/cri-api/pkg/streaming
```

### CODE REVIEW

File: staging/src/k8s.io/cri-api/go.mod [hunk 1]
```diff
@@ -7,17 +7,15 @@ module k8s.io/cri-api
...
 	github.com/stretchr/testify v1.2.2
 	golang.org/x/net v0.0.0-20190206173232-65e2d4e15006
-	golang.org/x/sync v0.0.0-20181108010431-42b317875d0f // indirect
-	golang.org/x/text v0.3.0 // indirect
-	google.golang.org/genproto v0.0.0-20170731182057-09f6ed296fc6 // indirect
 	google.golang.org/grpc v1.13.0
+	k8s.io/apimachinery v0.0.0
+	k8s.io/apiserver v0.0.0 // indirect
+	k8s.io/client-go v0.0.0
+	k8s.io/kubernetes v1.14.0
```
Comment (nikhita) [inline staging/src/k8s.io/cri-api/go.mod]: staging repos shouldn't be using k/k

Comment (nikhita) [reply staging/src/k8s.io/cri-api/go.mod]: https://github.com/kubernetes/kubernetes/blob/97011635f727307db25566a0459a51c6e1663836/staging/src/k8s.io/cri-api/pkg/streaming/server.go#L38

Comment (liggitt) [inline staging/src/k8s.io/cri-api/go.mod]: things in staging cannot depend on `k8s.io/kubernetes`

**Comment by liggitt** [comment]
looking through this, do we want the api and streaming implementation combined into a single module? the added import tree is significant for someone who just wanted the proto definitions

File: staging/src/k8s.io/cri-api/go.mod (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/cri-api/go.mod]: also:
https://github.com/kubernetes/kubernetes/blob/97011635f727307db25566a0459a51c6e1663836/staging/src/k8s.io/cri-api/pkg/streaming/server.go#L39

**Comment by alok87** [comment]
@nikhita @liggitt thanks for the review.

Opening a KEP for figuring out the location for KEP.

I saw few more dependencies after I moved `portforward` and `remotecommand`. 
```
./pkg/streaming/server_test.go:	api "k8s.io/kubernetes/pkg/apis/core"
./pkg/streaming/portforward/httpstream.go:	api "k8s.io/kubernetes/pkg/apis/core"
./pkg/streaming/portforward/httpstream_test.go:	api "k8s.io/kubernetes/pkg/apis/core"
./pkg/streaming/portforward/websocket.go:	api "k8s.io/kubernetes/pkg/apis/core"
./pkg/streaming/remotecommand/httpstream.go:	api "k8s.io/kubernetes/pkg/apis/core"
```

Any suggestion on isolating these dependencies? @dims @nikhita @liggitt

**Comment by alok87** [comment]
/hold 
a KEP is being raise to discuss on where should be the location of streaming library
- `staging/cri-api/pkg/streaming`
- `staging/cri-server`
- `staging/cri-helpers` or with some other name

File: staging/publishing/rules.yaml [hunk 2]
```diff
@@ -322,4 +322,13 @@ rules:
   - source:
       branch: master
       dir: staging/src/k8s.io/cri-api
+    dependencies:
```
Comment (nikhita) [inline staging/publishing/rules.yaml]: Can you move `name: master` above dependencies?

Sorry for not pointing this out earlier, but you could also add import restrictions for cri-api here: https://github.com/kubernetes/kubernetes/blob/master/staging/publishing/import-restrictions.yaml? Thanks!

Comment (alok87) [reply staging/publishing/rules.yaml]: Move the `name: master`
Added import restrictions based on https://github.com/alok87/kubernetes/blob/dep-streaming-moved-out/staging/src/k8s.io/cri-api/go.mod#L15-L20

Why do we specify import restrictions? What does the bot do reading this file?

Comment (nikhita) [reply staging/publishing/rules.yaml]: > Added import restrictions based on https://github.com/alok87/kubernetes/blob/dep-streaming-moved-out/staging/src/k8s.io/cri-api/go.mod#L15-L20

Ah, I meant adding the import restrictions to this file: https://github.com/kubernetes/kubernetes/blob/master/staging/publishing/import-restrictions.yaml.

> Why do we specify import restrictions? What does the bot do reading this file?

It makes sure that a package won't import any additional packages accidentally. Importing a new package would mean explicitly adding it in the list in that file.

The main significance of that file is to make sure that a staging repo won't import k/k though...

Comment (alok87) [reply staging/publishing/rules.yaml]: Got it. thanks
I have added it in `import-restrictions.yaml` only

Comment (nikhita) [reply staging/publishing/rules.yaml]: @alok87 I can't see the change in the diff. Maybe forgot to push? :upside_down_face:

Comment (alok87) [reply staging/publishing/rules.yaml]: please check now :)

**Comment by alok87** [comment]
@nikhita Can you help me figure out the reason for bazel verification failure. Update bazel script is working fine locally. Also what is bazel build, any doc link could help.

**Comment by nikhita** [comment]
@alok87 can you rebase and retry using `hack/update-bazel.sh`? If that doesn't work, let us know :)

**Comment by dims** [comment]
/close

let's revisit this after the KEP merges

**Pull request closed without merging**